### PR TITLE
[KNI] introduce the KNIDebug plugin

### PR DESCRIPTION
**Pull request by ffromani**
The KNIDebug is meant to be a passthrough plugin (which will fail only in the most catastrophic of cases) which we can use
to inject extra logs u/s is not willing or interesting to accept.

Using the KNIDebug plugin, we can choose between the tradeoff of changing the u/s code (which increases the maintainership
costs) at the cost of limiting ourselves in the logs we can add: some logs can only meaningfully by added in the private function
of some plugins.

Signed-off-by: Francesco Romani <fromani@redhat.com>

### CODE REVIEW

**Comment by ffromani** [comment]
example run:
```
$ oc logs -n numaresources secondary-scheduler-69f69d9bc8-fgnk5 | grep -i knidebug
I0524 09:00:23.402885       1 knidebug.go:50] "Creating new KNIDebug plugin"
      - name: KNIDebug
I0524 09:01:59.090687       1 knidebug.go:127] "target resource requests" logKey="e2e-test-workload-placement-resources-1919/testpod" cpu="16000 (16)" memory="17179869184 (16 GiB)"
I0524 09:01:59.090702       1 knidebug.go:143] "enough node resources" logKey="e2e-test-workload-placement-resources-1919/testpod" node="node4"
I0524 09:01:59.091040       1 knidebug.go:127] "target resource requests" logKey="e2e-test-workload-placement-resources-1919/testpod" cpu="16000 (16)" memory="17179869184 (16 GiB)"
I0524 09:01:59.091063       1 knidebug.go:143] "enough node resources" logKey="e2e-test-workload-placement-resources-1919/testpod" node="node5"
```

File: pkg-kni/knidebug/knidebug.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+	logKey := fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)
+	// note the fit.go plugin computes this in the prefilter stage. Does this make any practical difference in our context?
+	req := computePodResourceRequest(pod)
+	checkRequest(logKey, req, nodeInfo)
+	return nil
+}
+
+/////////////////////////////////////////////////////////////////////
+// taken from fit.go
+
```
Comment (swatisehgal) [inline pkg-kni/knidebug/knidebug.go:117]: Let's add a link here as well: https://github.com/kubernetes/kubernetes/blob/v1.24.0/pkg/scheduler/framework/plugins/noderesources/fit.go#L133-L175.

**Review by swatisehgal — COMMENTED** [review]
Overall, I understand the intent and reasoning behind the choice made in this PR to duplicate the filtering logic implemented as part of `NodeResourcesFit` plugin. It provides us a guard rail against any sub optimal scheduler configuration and we ensure that nodes that cannot fit a pod are filtered out. But, it appears to me that we are recommending that the scheduling profile should be configured such that KNIDebug plugin runs prior to `NodeResourceTopologyMatch` plugin as opposed to `NodeResourcesFit`. Am I understanding this correctly?

Rather than re-implementing  `NodeResourcesFit` filtering logic in this plugin, I was expecting that KNIDebug plugin would inform whether or not the scheduler was configured with `NodeResourcesFit` to run at pre-filtering stage. The key question here is it possible for a plugin to access the scheduler config or to identify other plugins that have been configured to run prior to it?

If the above is not possible, I am not entirely convinced that we use this new plugin to filter out nodes as opposed to emphasizing in the documentation and updating our scheduler config to make sure that `NodeResourcesFit` plugin is configured to run (at pre-filter stage) prior to `NodeResourceTopologyMatch`. The purpose of KNIDebug plugin should be to debug and gain visibility rather than automatically abstracting away from non-ideal configuration and taking care of filtering logic in its own hands.

File: pkg-kni/knidebug/knidebug.go (hunk 1 above)
Comment (ffromani) [reply pkg-kni/knidebug/knidebug.go:117]: thanks, fixed

**Comment by ffromani** [comment]
> Overall, I understand the intent and reasoning behind the choice made in this PR to duplicate the filtering logic implemented as part of `NodeResourcesFit` plugin. It provides us a guard rail against any sub optimal scheduler configuration and we ensure that nodes that cannot fit a pod are filtered out. But, it appears to me that we are recommending that the scheduling profile should be configured such that KNIDebug plugin runs prior to `NodeResourceTopologyMatch` plugin as opposed to `NodeResourcesFit`. Am I understanding this correctly?

For this filter to be useful, it needs to run before (but not necessarily **right before**) the noderesourcefit plugin.
In turn, it's probably better to run noderesourcefit before numatopologymatch, even though this is somehow debatable.

> > Rather than re-implementing `NodeResourcesFit` filtering logic in this plugin, I was expecting that KNIDebug plugin would inform whether or not the scheduler was configured with `NodeResourcesFit` to run at pre-filtering stage. The key question here is it possible for a plugin to access the scheduler config or to identify other plugins that have been configured to run prior to it?

I think it is doable, but doesn't seem trivial. However I'm not sure about what do we gain in this case?

> If the above is not possible, I am not entirely convinced that we use this new plugin to filter out nodes as opposed to emphasizing in the documentation and updating our scheduler config to make sure that `NodeResourcesFit` plugin is configured to run (at pre-filter stage) prior to `NodeResourceTopologyMatch`.

In the default configuration this is the case. The scheduler framework docs is not crystal clear in this case, but it seems (and we should probably doublecheck) that explicitely enable filters are appended at the end of the default chain.

> The purpose of KNIDebug plugin should be to debug and gain visibility rather than automatically abstracting away from non-ideal configuration and taking care of filtering logic in its own hands.

I fully agree. But I'm not clear what in the current implementation goes against this direction. The entirety of the `knidebug` plugin can be proposed as noderesourcefit patch - it only add logs after all. Thing is, I tend to believe upstream would not like such a change too much, considering the noderesourcefit is pretty stable and well understood and that we can have the same result with this out of tree implementation.

**Comment by swatisehgal** [comment]
Ah!! I believe I might have misunderstood the goals of this PR. I got an impression that KNIDebug will be used as a replacement of `NodeResourcesFit` plugin and perform the filtering of nodes as done by `NodeResourcesFit`. That is my bad, I misread the filter function!

I believe, we are saying here that this KNIDebug plugin runs prior to the `NodeResourceTopologyMatch` plugin and ideally after `NodeResourcesFit` plugin where nodes that cannot fit pods have ideally already been filtered out but in scenarios that `NodeResourcesFit` hasn't been configured we still have logs indicating that the node being considered for scheduling does not have enough resources. Is this a fair summarization of the goal here?

Just a couple of questions because I haven't tested these out and haven't found documentation on this, so asking in case you have already played around with this:

1. Can we control the order of plugin execution? Is it directly linked to the order in which they are configured in the scheduler configuration?
2. Does the scheduler framework allow us to mix and match custom plugins (`NodeResourceTopologyMatch` and `KNIDebug` here) with a plugin part of default scheduler profile (in this case `NodeResourcesFit`) ?

**Comment by ffromani** [comment]
> I believe, we are saying here that this KNIDebug plugin runs prior to the `NodeResourceTopologyMatch` plugin and ideally after `NodeResourcesFit` plugin where nodes that cannot fit pods have ideally already been filtered out but in scenarios that `NodeResourcesFit` hasn't been configured we still have logs indicating that the node being considered for scheduling does not have enough resources. Is this a fair summarization of the goal here?

Almost there. The goal is to give more visibility into the decision taken by `NodeResourcesFit` to easily understand why a node was discarded (with more details) and to easily evaluate the interplay with `NodeTopologyMatch`. This is the counterpart to the logging enhancements we recently sent to the scheduler plugins repo, we want to have a similar experience with this plugin to
be able to easily debug and troubleshoot our flow.

And since this is actually about our flow, this is another reason to believe u/s would be really tiepid about this change - but that's no biggie after all, and pretty fair in the end.

>     1. Can we control the order of plugin execution? Is it directly linked to the order in which they are configured in the scheduler configuration?

Yes to both. The downside is we need to explicitely list all the plugins we want (the default is "everything you have, whatever it is") which means we will need to review the list when a new version comes out (e.g. 1.23->1.24 and so on). I think this is still a pretty fair price to pay.

>     2. Does the scheduler framework allow us to mix and match custom plugins (`NodeResourceTopologyMatch` and `KNIDebug` here) with a plugin part of default scheduler profile (in this case `NodeResourcesFit`) ?

yes, there should be no difference after plugins are compiled in the secondary scheduler

**Comment by ffromani** [comment]
/hold
let's add some tests

**Comment by swatisehgal** [comment]
Thanks for the clarification. I am happy enough to proceed in this direction.

I will take another pass once we have tests but so far the PR looks good to me.

**Comment by ffromani** [comment]
/hold cancel
I believe the test coverage is sufficient

**Comment by Tal-or** [comment]
Very nice. Are we going to have this plugin enabled by default?

**Comment by ffromani** [comment]
> Very nice. Are we going to have this plugin enabled by default?

yes, because we will control the logs using the `LogLevel` tunable

**Pull request merged**